tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,423 @@
+
+        Parameters
+        ----------
+        phi_block : DataFrame
+            Saha Factors for current atomic number
+        electron_density : float
+            Current guess for electron density
+
+        Returns
+        -------
+        numpy.array
+            LTE block for rate matrix
+        """
+        lte_rate_vector_block = -1.0 * np.hstack([*phi_block.values, -1.0])
+        lte_rate_matrix_block = np.diag(lte_rate_vector_block)
+        n_e_initial = np.ones(len(phi_block)) * electron_density
+        n_e_matrix = np.diag(n_e_initial, 1)
+        lte_rate_matrix_block += n_e_matrix
+        lte_rate_matrix_block[-1, :] = 1.0
+        return lte_rate_matrix_block
+
+    @staticmethod
+    def prepare_phi(phi):
+        """
+        Makes sure that phi does not have any 0 entries.
+        """
+        phi[phi == 0.0] = 1.0e-10 * phi[phi > 0.0].min().min()
+        return phi
+
+    @staticmethod
+    def recomb_matrix(recomb_coefficients, atomic_number, ion_number):
+        """Constructs a recombination rate matrix from the recombination rates.
+
+        Parameters
+        ----------
+        recomb_coefficients : DataFrame
+            Recombination coefficients.
+        atomic_number : int64
+            Current atomic number. Used for the dimension of a square matrix.
+        ion_number : int64
+            Current ion number. Used for returning the correct row.
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        offdiag = np.zeros(atomic_number)
+        index = recomb_coefficients.index
+        for i in index:
+            offdiag[i] = recomb_coefficients.loc[i]
+        diag = np.hstack([np.zeros(1), -offdiag])
+        return (np.diag(diag) + np.diag(offdiag, k=1))[ion_number, :]
+
+    @staticmethod
+    def ion_matrix(ion_coefficients, atomic_number, ion_number):
+        """Constructs an ionization rate matrix from the ionization rates.
+
+        Parameters
+        ----------
+        ion_coefficients : DataFrame
+            Recombination coefficients.
+        atomic_number : int64
+            Current atomic number. Used for the dimension of a square matrix.
+        ion_number : int64
+            Current ion number. Used for returning the correct row.
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        offdiag = np.zeros(atomic_number)
+        index = ion_coefficients.index
+        for i in index:
+            offdiag[i] = ion_coefficients.loc[i]
+        diag = np.hstack([-offdiag, np.zeros(1)])
+        return (np.diag(diag) + np.diag(offdiag, k=-1))[ion_number, :]
+
+    @staticmethod
+    def prepare_charge_conservation_row(atomic_numbers):
+        """Prepares the last row of the rate_matrix. This row corresponds to the charge
+        density equation."""
+        charge_conservation_row = []
+        for atomic_number in atomic_numbers:
+            charge_conservation_row.append(np.arange(0.0, atomic_number + 1))
```

Comment:
```suggestion
            charge_conservation_row.append(np.arange(0, atomic_number + 1))
```
@chvogl is this fine?